src/guide/customization.rs

```diff
@@ -0,0 +1,398 @@
+//! A tutorial on how to customize eyre by overriding the Context parameter
+//!
+//! # Introduction
+//!
+//! In this tutorial we will go through the process of implementing a custom context step by step:
+//!
+//! * Start with defining an empty context that captures nothing
+//! * Add support for capturing backtraces
+//!     * skip backtrace capture if our source captured one
+//! * Add support for a http status codes
+//!     * Implement a setter for Result<T, Report> with a new Trait
+//!
+//! # A Basic Context
+//!
+//! The simplest context is an empty context, so lets start there, here's what our context looks
+//! like:
+//!
+//! ```rust
+//! struct CustomContext;
+//! ```
+//!
+//! Now this context doesn't add much context, but it does have the advantage of removing context,
+//! aka the backtrace, that may be unnecessary. To use this context we need to implement
+//! `EyreContext`. This provides the necessary interface for eyre to construct and use your context.
+//!
+//! First, we have to impl `default` so eyre knows how to construct your context when it creates
+//! your `Report` from a `std::error::Error`:
+//!
+//!
+//! ```rust
+//! fn default(_: &(dyn std::error::Error + 'static)) -> Self {
+//!     Self
+//! }
+//! ```
+//!
+//! And next we need to implement `debug` to tell eyre how to format the final report including
+//! your custom context:
+//!
+//! ```rust
+//! use eyre::{Chain, Indenter};
+//!
+//! fn debug(
+//!     &self,
+//!     error: &(dyn std::error::Error + 'static),
+//!     f: &mut core::fmt::Formatter<'_>,
+//! ) -> core::fmt::Result {
+//!     if f.alternate() {
+//!         return core::fmt::Debug::fmt(error, f);
+//!     }
+//!
+//!     let errors = Chain::new(error).enumerate();
+//!
+//!     for (n, error) in errors {
+//!         writeln!(f)?;
+//!         write!(Indented::numbered(f, n), "{}", error)?;
+//!     }
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! Things to note, we're using some utilities provided by `eyre` for writing error reports,
+//!
+//! - `Chain`: iterator for errors while the `chain` fn on the error trait is still unstable.
+//! - `Indenter`: a wrapper for formatters that inserts indentation after newlines for handling
+//! multi line error messages nicely
```

Comment:
This is a bit of an awkward place to mention the crates being used. I think it could perhaps be nice to mention them off-hand at the beginning, and have more detailed information in an appendix-type thing?